src/java.base/share/classes/sun/security/ssl/DHasKEM.java

```diff
@@ -0,0 +1,354 @@
+        public int engineSecretSize() {
+            return params.secretLen;
+        }
+
+        @Override
+        public int engineEncapsulationSize() {
+            return params.publicKeyLen;
+        }
+
+        @Override
+        public SecretKey engineDecapsulate(byte[] encapsulation, int from,
+                int to, String algorithm) throws DecapsulateException {
+            if (encapsulation.length != params.publicKeyLen) {
+                throw new DecapsulateException("incorrect encapsulation size");
+            }
+            try {
+                PublicKey pkE = params.DeserializePublicKey(encapsulation);
+                SecretKey dh = params.DH(algorithm, skR, pkE);
+                return sub(dh, from, to);
+            } catch (IOException | InvalidKeyException e) {
+                throw new DecapsulateException("Cannot decapsulate", e);
+            } catch (Exception e) {
+                throw new ProviderException("internal error", e);
+            }
+        }
+
+        private SecretKey sub(SecretKey key, int from, int to) {
+            if (from == 0 && to == params.secretLen) {
+                return key;
+            }
+
+            // Key slicing should never happen. Otherwise, there might be
+            // a programming error.
+            throw new AssertionError(
+                    "Unexpected key slicing: from=" + from + ", to=" + to);
+        }
+
+        // This KEM is designed to be able to represent every ECDH and XDH
+        private Params paramsFromKey(Key k) throws InvalidKeyException {
+            if (k instanceof ECKey eckey) {
+                if (ECUtil.equals(eckey.getParams(), CurveDB.P_256)) {
+                    return Params.P256;
+                } else if (ECUtil.equals(eckey.getParams(), CurveDB.P_384)) {
+                    return Params.P384;
+                } else if (ECUtil.equals(eckey.getParams(), CurveDB.P_521)) {
+                    return Params.P521;
+                }
+            } else if (k instanceof XECKey xkey
+                    && xkey.getParams() instanceof NamedParameterSpec ns) {
+                if (ns.getName().equalsIgnoreCase(
+                        NamedParameterSpec.X25519.getName())) {
+                    return Params.X25519;
+                } else if (ns.getName().equalsIgnoreCase(
+                        NamedParameterSpec.X448.getName())) {
+                    return Params.X448;
+                }
+            }
+            throw new InvalidKeyException("Unsupported key");
+        }
+    }
+
+    private enum Params {
+
+        P256(32, 2 * 32 + 1,
+                "ECDH", "EC", CurveDB.P_256),
+
+        P384(48, 2 * 48 + 1,
+                "ECDH", "EC", CurveDB.P_384),
+
+        P521(66, 2 * 66 + 1,
+                "ECDH", "EC", CurveDB.P_521),
+
+        X25519(32, 32,
+                "XDH", "XDH", NamedParameterSpec.X25519),
+
+        X448(56, 56,
+                "XDH", "XDH", NamedParameterSpec.X448),
```

Comment:
unneeded comma, move the semicolon to the end of the "XDH" line (currently 261).

Reply:
Removed.